Watches

samoht opened the pull request:
Not ready to merge yet. Here is the current status:

- the code compiles but is totally untested (so assume it is broken)

- there is no unit-tests at all for the notification code (the old one as well as the new one). I'll add extensive tests before envisaging merging.

- the new API is much simpler than the previous one, with explicit de-allocation and clearer concurrency semantics. Basically you now only have:
    ```ocaml
  type 'a diff = [ `Updated of 'a * 'a | `Added of 'a | `Removed of 'a ]

  val watch_head: t -> ?init:head -> (head diff -> unit Lwt.t) ->
    (unit -> unit Lwt.t) Lwt.t
  (** [watch_tag t f] calls [f] everytime the contents of [t]'s tag is
      updated. Do nothing if [t] is not persistent. Return a clean-up
      function to remove the watch handler.

      {b Note:} even [f] might skip some head updates, it will never
      be called concurrently: all consecutive calls to [f] are done in
      sequence, so we ensure that the previous one ended before
      calling the next one. *)

  val watch_tags: t -> ?init:(tag * head) list ->
    (tag -> head diff -> unit Lwt.t) -> (unit -> unit Lwt.t) Lwt.t
  (** [watch_tags t f] calls [f] to the watch handlers of {b all} tag
      changes in the local store. Return a function to remove the
      handler. *)
    ```
     I'm not totally convinced about the function names yet, but the types are ok now I think. Basically, you register a callback which is called on every update. If multiple callbacks are registered, they are called in sequence, and they won't overloap with each other. 

- There is not (yet?) an optimisation to clean-up the callback calls in case the waiting queue become too big and starts to lag. I'm not sure it is necessary, I plan to run some stress-tests to see how it works under load with lots of long-running callback function.

- The previous API which let you watch for changes in sub-path has been completely removed. This was designed first to look like xenstore watches, but with subtle semantic differences so not really useful as they were. I plan to maybe add back something similar later if people needs it, but for now I prefer relying on fewer well-tested functions.

yansh commented:
> The previous API which let you watch for changes in sub-path has been completely removed. 

Actually, I was wondering about that. Is there a way to watch changes to value in path, i.e., I commit new value?

samoht commented:
> Actually, I was wondering about that. Is there a way to watch changes to value in path, i.e., I commit new value?

not anymore, but I can add back that some helper functions doing that if it is useful for you.

yansh commented:
Yes, but let's see understand it correctly:

If I have a path ```/A/B/C```  with  ```value```; when I commit new value (e.g., ```value1```) to path it will trigger some function. 

If the described logic is correct, then yes please that would help a lot :) Why remove it in a first place, sounds like a very neat feature to have :)

samoht commented:
Previously, if you watched the path "/A/B/C" and you updated "/A/B/C" or "/A/B/C/<path>" with a new value, then you got a notification. The function was (i) broken and (ii) you got a lot of notification if you updated a with a view (as a lot of paths can change at the same time) (iii) you had no clear idea if the value you got is actually the most recent one (as you might receive an old notification which took time to be transmitted) and (iv) it wasn't possible to stop listening without leaking resources.

I'm trying to build the base functions to solve all of these first, and then I'll add back more complex watch functions :p

yansh commented:
Understood. Particularly the (ii) point. Yes, without solving these problems, I think such watch function is practically useless.

talex5 commented:
It seems odd to have diffs and skipping together in one API. I'd expect either:
- After a change, the callback is eventually called with the current state, possibly skipping some earlier states if busy, or
- The callback is called with every diff.

(I'd prefer the first)

With the proposed API, what happens if the updates are e.g.

1. `Added 1`
2. `Updated (1, 2)`

but we're busy and so only see (2)? Do we get `Updated (1, 2)` when we never saw the `Added (1)`? Or do we get `Added 2` (a diff that never actually ocurred)?

samoht commented:
@talex5 currently there is no skipping/coalescing, so the callback will be called on every diff. I was planning to add a coalescing step to have `Added 2`.  Note: on some backends (especially the one using active polling such as disk-based one) some updates might be skipped anyway so the client code should be able to handle that anyway.

samoht commented:
@dsheets any idea what's happening here: https://travis-ci.org/mirage/irmin/jobs/58355929#L1642

Seems that either uri or cohttp are confused with percent-encoded query strings.

dsheets commented:
I'm looking into it. Where are the URIs constructed/deconstructed?

dsheets commented:
P.S. alcotest 0.3.0 puts the stack traces into the output files but I guess Travis doesn't let you access those :-(

samoht commented:
I'm starting to be happy with the stability and quality of that code. In addition to `watch_head` and `watch_tags` I've also added:

```ocaml
(* In Irmin.S *)

  val watch_key: t -> key -> ?init:(head * value) ->
    ((head * value) diff -> unit Lwt.t) -> (unit -> unit Lwt.t) Lwt.t
  (** [watch_key t key f] calls [f] every time the [key]'s value is
      added, removed or updated. *)

(* in Irmin.VIEW *)

  val watch_path: db -> key -> ?init:(head * t) ->
    ((head * t) diff -> unit Lwt.t) -> (unit -> unit Lwt.t) Lwt.t
  (** [watch_head t p f] calls [f] every time subpaths of [p] are
      updated in the branch [t]. The callback parameters contains
      branch's current head and the corresponding view. *)
```

samoht commented:
Watche tests have been made more synchronous and now pass properly with various polling delay (and thus work now fine on Travis CI). Merging!
